backend/dataall/base/services/service_provider_factory.py

```diff
@@ -0,0 +1,15 @@
+import os
+
+from dataall.base.aws.cognito import Cognito
+
+
+class ServiceProviderFactory:
+
+    @staticmethod
+    def get_service_provider_instance():
+        if (os.environ.get("custom_auth", None)):
+            # Return instance of your service provider which implements the ServiceProvider interface
+            # Please take a look at the "Deploy to AWS" , External IDP section for steps
+            return None
```

Comment:
If a customer implements custom_auth, the step to update the ServiceProviderFactory is a deployment requirement.

In the case a customer forgets to update this line with their own ServiceProvider, it will only fail once the api is called, returning an error like "Cannot call get_groups_for_user from None"
"""
def get_custom_groups(user_id):
    service_provider = ServiceProviderFactory.get_service_provider_instance()
    return service_provider.get_groups_for_user(user_id)
"""


To avoid this situation we could add more guardrails:
1. In this file, instead of returning None, we can raise and Exception with a descriptive error log
2. At the time of deployment (in the deploy/stack/pipeline.py), we can validate the `service_provider = ServiceProviderFactory.get_service_provider_instance()`

What do you think?

Reply:
I think having a guardrail for this would be very helpful. 

For 1. I agree I will raise an exception instead of returning None. Thus, the logs can clearly state what went wrong. 

For 2. I don't think so this would be easily doable as the code depends on setting of an environment variable `custom_auth` which is set in the lambda code. 

In order to do such a thing, while in the backend stack, this environment variable has to be set and then an instance of service_provider could be created and checked like 
```python
        if custom_auth is not None:
            try:
                os.environ['custom_auth'] = 'externalIDP'
                ServiceProviderFactory.get_service_provider_instance()
            except Exception as e:
                if str(e) == 'Service Provider not implemented when using custom auth configuration. Please implement and the again deploy backend stack':
                    del os.environ['custom_auth']
                    raise Exception('Service Provider not implemented when using custom auth configuration. Please implement and the again deploy backend stack')
```

But in order for the `ServiceProviderFactory.get_service_provider_instance()` to be imported in the `deploy` the `backend` folder has to be made package with an `__init__.py` file exporting the `dataall` package. 

Please let me know if there have another better way to do this check. Also, what do you think of the above way of adding this guardrail

Reply:
I overlooked the os.environ. mmm I see that it makes it harder to implement. In that case I think the error message and the documentation will be enough.